File: luogu/P4924.py
def rotate(x, y, r, dir):
    global map_, right_left, bottom_up
    if r == 0:
        return

    for i in range(2):
        for j in range(2*r+1):
            c = 1 if i == 0 else -1
            right_left[i][j] = map_[x - r + j][y + c*r]
            bottom_up[i][j] = map_[x + c*r][y - r + j]

    if dir == 0:
        for i in range(2*r + 1):
            map_[x - r][y - r + i] = right_left[1][2*r - i]
        for i in range(2*r + 1):
            map_[x - r + i][y + r] = bottom_up[1][i]
        for i in range(2*r + 1):
            map_[x + r][y + r - i] = right_left[0][i]
        for i in range(2*r + 1):
            map_[x + r - i][y - r] = bottom_up[0][2*r - i]
    else:
        for i in range(2*r + 1):
            map_[x - r][y - r + i] = right_left[0][i]
        for i in range(2*r + 1):
            map_[x - r + i][y + r] = bottom_up[0][2*r - i]
        for i in range(2*r + 1):
            map_[x + r][y + r - i] = right_left[1][2*r - i]
        for i in range(2*r + 1):
            map_[x + r - i][y - r] = bottom_up[1][i]

    rotate(x, y, r - 1, dir)


if __name__ == "__main__":
    n, m = map(int, input().split())

    map_ = [[0] * (n + 1) for _ in range(n + 1)]
    # 从上到下拷贝
    right_left = [[0] * 505 for _ in range(2)]
    # 从左到右拷贝
    bottom_up = [[0] * 505 for _ in range(2)]

    temp = 1

    for i in range(1, n + 1):
        for j in range(1, n + 1):
            map_[i][j] = temp
            temp += 1

    for _ in range(m):
        x, y, r, z = map(int, input().split())
        rotate(x, y, r, z)

    for i in range(1, n + 1):
        for j in range(1, n + 1):
            print(map_[i][j], end=" ")
        print()

File: tools/records/P4924.cpp
#include <iostream>

int map[505][505];
int right_left[2][505];
int bottom_up[2][505];
int n, m;

void rotate(int x, int y, int r, int dir) {
    if (r == 0) return;

    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 2 * r + 1; j++) {
            int c = (i == 0 ? 1 : -1);
            right_left[i][j] = map[x - r + j][y + c*r];
            bottom_up[i][j] = map[x + c*r][y - r + j];
        }
    }

    if (dir == 0) {
        for (int i = 0; i < 2*r + 1; i++)
            map[x - r][y - r + i] = right_left[1][2*r - i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x - r + i][y + r] = bottom_up[1][i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x + r][y + r - i] = right_left[0][i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x + r - i][y - r] = bottom_up[0][2*r - i];
    } else {
        for (int i = 0; i < 2*r + 1; i++)
            map[x - r][y - r + i] = right_left[0][i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x - r + i][y + r] = bottom_up[0][2*r - i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x + r][y + r - i] = right_left[1][2*r - i];
        for (int i = 0; i < 2*r + 1; i++)
            map[x + r - i][y - r] = bottom_up[1][i];
    }
        
    rotate(x, y, r - 1, dir);
}

int main() {
    std::ios_base::sync_with_stdio(false);

    std::cin >> n >> m;
    int temp = 1;

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            map[i][j] = temp++;
        }
    }

    for (int i = 0; i < m; i++) {
        int x, y, r, z;
        std::cin >> x >> y >> r >> z;

        rotate(x, y, r, z);
    }

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            std::cout << map[i][j] << " ";
        }
        std::cout << "\n";
    }
}
